modules/ximgproc/src/scansegment.cpp

```diff
@@ -0,0 +1,926 @@
+				offsetVec[i] = static_cast<int*>(malloc(offsetSize));
+				if (offsetVec[i] == NULL) {
+					offsetAllocated = false;
+				}
+			}
+			for (int i = 0; i < neighbourCount; i++) {
+				neighbourLocBuffer[i] = (neighbourLoc[i].y * width) + neighbourLoc[i].x;
+			}
+
+			if (labelsBuffer != NULL && clusterBuffer != NULL && pixelBuffer != NULL && offsetAllocated) {
+				setupComplete = true;
+			}
+			else {
+				setupComplete = false;
+
+				if (labelsBuffer == NULL) {
+					CV_Error(Error::StsInternal, "Cannot initialise labels buffer");
+				}
+				if (clusterBuffer == NULL) {
+					CV_Error(Error::StsInternal, "Cannot initialise cluster buffer");
+				}
+				if (pixelBuffer == NULL) {
+					CV_Error(Error::StsInternal, "Cannot initialise pixel buffer");
+				}
+				if (!offsetAllocated) {
+					CV_Error(Error::StsInternal, "Cannot initialise offset buffers");
+				}
+			}
+		}
+
+		ScanSegmentImpl::~ScanSegmentImpl()
+		{
+			// clean up
+			if (neighbourLoc != NULL) {
+				free(neighbourLoc);
+			}
+			if (seedRects != NULL) {
+				free(seedRects);
+			}
+			if (seedRectsExt != NULL) {
+				free(seedRectsExt);
+			}
+			if (offsetRects != NULL) {
+				free(offsetRects);
+			}
+			if (labelsBuffer != NULL) {
+				free(labelsBuffer);
+			}
+			if (clusterBuffer != NULL) {
+				free(clusterBuffer);
+			}
+			if (pixelBuffer != NULL) {
+				free(pixelBuffer);
+			}
+			for (int i = 0; i < effectivethreads; i++) {
+				if (offsetVec[i] != NULL) {
+					free(offsetVec[i]);
+				}
+			}
+			if (!src.empty()) {
+				src.release();
+			}
+			if (!labelsMat.empty()) {
+				labelsMat.release();
+			}
+		}
+
+		void ScanSegmentImpl::iterate(InputArray img)
+		{
+			// ensure setup successfully completed
+			CV_Assert(setupComplete);
+
+			if (img.isMat())
+			{
+				// get Mat
+				src = img.getMat();
+
+				// image should be valid
+				CV_Assert(!src.empty());
+			}
+			else if (img.isMatVector())
+			{
+				std::vector<cv::Mat> vec;
+
+				// get vector Mat
+				img.getMatVector(vec);
+
+				// array should be valid
+				CV_Assert(!vec.empty());
+
+				// merge into Mat
+				cv::merge(vec, src);
+			}
+			else
+				CV_Error(Error::StsInternal, "Invalid InputArray.");
+
+			int depth = src.depth();
+
+			CV_Assert(src.size().width == width && src.size().height == height);
+			CV_Assert(depth == CV_8U);
+			CV_Assert(src.channels() == 3);
+
+			clusterCount = 0;
+			clusterIndex.store(0);
+			clusterID.store(1);
+
+			smallClusters = indexSize / smallClustersDiv;
+
+			// set labels to NONE
+			labelsMat.setTo(NONE);
+
+			// set labels buffer to UNCLASSIFIED
+			std::fill(labelsBuffer, labelsBuffer + indexSize, UNCLASSIFIED);
+
+			// apply light blur
+			cv::medianBlur(src, src, 3);
+
+			// start at the center of the rect, then run through the remainder
+			labBuffer = reinterpret_cast<cv::Vec3b*>(src.data);
+			cv::parallel_for_(cv::Range(0, (int)indexNeighbourVec.size()), PP1(reinterpret_cast<ScanSegmentImpl*>(this)));
```

Comment:
BTW, C++11 allows to use lambdas here instead of manual ParallelLoopBody classes.

Reply:
I would prefer to leave this unchanged as my editor (Visual Studio Community) can't compile C++ 11 directly. I can only set it to ISO C++ 14 or 17, so changing to lambdas may introduce errors which I can't check. Unless someone else can help :)